File: sdk/python/core/keeper_secrets_manager_core/dto/payload.py
# -*- coding: utf-8 -*-
#  _  __
# | |/ /___ ___ _ __  ___ _ _ (R)
# | ' </ -_) -_) '_ \/ -_) '_|
# |_|\_\___\___| .__/\___|_|
#              |_|
#
# Keeper Secrets Manager
# Copyright 2023 Keeper Security Inc.
# Contact: sm@keepersecurity.com

class Context:

    def __init__(self, transmission_key, client_id, client_key):
        self.transmissionKey = transmission_key
        self.clientId = client_id
        self.clientKey = client_key


class TransmissionKey:

    def __init__(self, publicKeyId, key, encryptedKey):
        self.publicKeyId = publicKeyId
        self.key = key
        self.encryptedKey = encryptedKey


class GetPayload:

    def __init__(self):
        self.clientVersion = None
        self.clientId = None
        self.publicKey = None
        self.requestedRecords = None
        self.requestedFolders = None


class CreatePayload:

    def __init__(self):
        self.clientVersion = None
        self.clientId = None
        self.recordUid = None
        self.recordKey = None
        self.folderUid = None
        self.folderKey = None
        self.data = None
        self.subFolderUid = None


class DeletePayload:
    def __init__(self):
        self.clientVersion = None
        self.clientId = None
        self.recordUids = None


class CreateFolderPayload:
    def __init__(self):
        self.clientVersion = None
        self.clientId = None
        self.folderUid = None
        self.sharedFolderUid = None
        self.sharedFolderKey = None
        self.data = None
        self.parentUid = None


class UpdateFolderPayload:
    def __init__(self):
        self.clientVersion = None
        self.clientId = None
        self.folderUid = None
        self.data = None


class DeleteFolderPayload:
    def __init__(self):
        self.clientVersion = None
        self.clientId = None
        self.folderUids = None
        self.forceDeletion = False


class FileUploadPayload:

    def __init__(self):
        self.clientVersion = None
        self.clientId = None
        self.fileRecordUid = None
        self.fileRecordKey = None
        self.fileRecordData = None
        self.ownerRecordUid = None
        self.ownerRecordData = None
        self.linkKey = None
        self.fileSize = None


class UpdatePayload:

    def __init__(self):
        self.clientVersion = None
        self.clientId = None
        self.recordUid = None
        self.data = None
        self.revision = None
        self.transactionType = None  # 'general' or 'rotation'


class CompleteTransactionPayload:

    def __init__(self):
        self.clientVersion = None
        self.clientId = None
        self.recordUid = None


class EncryptedPayload:

    def __init__(self, encrypted_payload, signature):
        self.encrypted_payload = encrypted_payload
        self.signature = signature


class KSMHttpResponse:

    def __init__(self, status_code, data, http_response=None):
        self.status_code = status_code
        self.data = data
        self.http_response = http_response


class QueryOptions:

    def __init__(self, records_filter, folders_filter):
        self.records_filter = records_filter
        self.folders_filter = folders_filter


class CreateOptions:

    def __init__(self, folder_uid, subfolder_uid):
        self.folder_uid = folder_uid
        self.subfolder_uid = subfolder_uid

File: sdk/ruby/lib/keeper_secrets_manager/dto/payload.rb
module KeeperSecretsManager
  module Dto
    # Transmission key for encrypted communication
    class TransmissionKey
      attr_accessor :public_key_id, :key, :encrypted_key

      def initialize(public_key_id:, key:, encrypted_key:)
        @public_key_id = public_key_id
        @key = key
        @encrypted_key = encrypted_key
      end
    end

    # Base payload class
    class BasePayload
      attr_accessor :client_version, :client_id

      def to_h
        hash = {}
        instance_variables.each do |var|
          key = var.to_s.delete('@')
          value = instance_variable_get(var)

          # Convert Ruby snake_case to camelCase for API
          api_key = Utils.snake_to_camel(key)
          hash[api_key] = value unless value.nil?
        end
        hash
      end

      def to_json(*args)
        to_h.to_json(*args)
      end
    end

    # Get secrets payload
    class GetPayload < BasePayload
      attr_accessor :public_key, :requested_records, :requested_folders, :file_uids, :request_links

      def initialize
        super()
        @requested_records = nil
        @requested_folders = nil
        @file_uids = nil
        @request_links = nil
      end
    end

    # Create record payload
    class CreatePayload < BasePayload
      attr_accessor :record_uid, :record_key, :folder_uid, :folder_key,
                    :data, :sub_folder_uid

      def initialize
        super()
      end
    end

    # Update record payload
    class UpdatePayload < BasePayload
      attr_accessor :record_uid, :data, :revision, :transaction_type, :links2_remove

      def initialize
        super()
        @transaction_type = 'general'
      end
    end

    # Delete records payload
    class DeletePayload < BasePayload
      attr_accessor :record_uids

      def initialize
        super()
        @record_uids = []
      end
    end

    # Complete transaction payload
    class CompleteTransactionPayload < BasePayload
      attr_accessor :record_uid

      def initialize
        super()
      end
    end

    # File upload payload
    class FileUploadPayload < BasePayload
      attr_accessor :file_record_uid, :file_record_key, :file_record_data,
                    :owner_record_uid, :owner_record_data, :owner_record_revision, :link_key, :file_size

      def initialize
        super()
      end
    end

    # Create folder payload
    class CreateFolderPayload < BasePayload
      attr_accessor :folder_uid, :shared_folder_uid, :shared_folder_key,
                    :data, :parent_uid

      def initialize
        super()
      end
    end

    # Update folder payload
    class UpdateFolderPayload < BasePayload
      attr_accessor :folder_uid, :data

      def initialize
        super()
      end
    end

    # Delete folder payload
    class DeleteFolderPayload < BasePayload
      attr_accessor :folder_uids, :force_deletion

      def initialize
        super()
        @folder_uids = []
        @force_deletion = false
      end
    end

    # Encrypted payload wrapper
    class EncryptedPayload
      attr_accessor :encrypted_payload, :signature

      def initialize(encrypted_payload:, signature:)
        @encrypted_payload = encrypted_payload
        @signature = signature
      end
    end

    # HTTP response wrapper
    class KSMHttpResponse
      attr_accessor :status_code, :data, :http_response

      def initialize(status_code:, data:, http_response: nil)
        @status_code = status_code
        @data = data
        @http_response = http_response
      end

      def success?
        status_code >= 200 && status_code < 300
      end
    end
  end
end
